### PR TITLE
Iap iOS add payment dart wrappers

### PR DESCRIPTION
**Pull request by cyanglaz**
[iOS]
Dart wrappers for the required objects that are used when adding a payment. 
The code for adding a payment will be in a following PR after this one if approved and merged. 
https://github.com/flutter/flutter/issues/26328

### CODE REVIEW

File: packages/in_app_purchase/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart [hunk 1]
```diff
@@ -3,12 +3,280 @@
...
+
+  /// The app canceled the download.
+  @JsonValue(5)
+  cancelled,
+}
+
+/// Dart wrapper around StoreKit's [SKDownload](https://developer.apple.com/documentation/storekit/skdownload?language=objc).
+///
+/// When a product is created in the App Store Connect, one or more download contents can be associated with it.
+/// When the product is purchased, a List of download object will present in an [SKPaymentTransactionWrapper] object.
```
Comment (mklim) [inline packages/in_app_purchase/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart]: nit: "download object will present in an [SKPaymentTransactionWrapper] object" is a little hard to parse. I think it's mainly because [when "present" is used this way](https://www.merriam-webster.com/dictionary/present) it's a [transitive verb](https://en.wikipedia.org/wiki/Transitive_verb#In_English). Transitive verbs need an object to make sense, and it's missing one here. "download object" should probably also be plural since there's multiple in a List.

Maybe consider rewording to "[SkPaymentTransactionWrapper] will return a list of downloads"?

File: packages/in_app_purchase/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart [hunk 2]
```diff
@@ -3,12 +3,280 @@
...
+/// Then download objects can be added to the payment queue.
+/// Read the [contentURL] to get the URL of the downloaded content after download complete.A
```
Comment (mklim) [inline packages/in_app_purchase/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart]: nit: "after the download completes."

"download" needs a definite article, and "to complete" needs to be in the singular form of "to complete" [so that it matches the number of downloads](https://en.wikipedia.org/wiki/Agreement_(linguistics)#Number).

File: packages/in_app_purchase/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart [hunk 3]
```diff
@@ -3,12 +3,280 @@
...
+/// All downloaded files must be processed before the completion of the transaction.
+/// After the transaction is complete, any download object in the transaction will not be able to add to the payment queue
```
Comment (mklim) [inline packages/in_app_purchase/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart]: Is "download object" talking about a type of object that has the property of being able to be downloaded? Or is it talking about an object that has been downloaded in the past?

I'm asking because this wording is hard to parse as it is. "downloadable object" would be using "download" in the form of an adjective, and saying that the object has the quality of being able to be downloaded. "downloaded object" would be using "download" as a [past participle](https://en.wikipedia.org/wiki/Participle#Types_of_participle), and saying that the objects have definitely been downloaded at some prior point.

Comment (mklim) [inline packages/in_app_purchase/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart]: nit: "will not be able to add to the payment queue" -> "will not be able to be added to the payment queue"

I'm having trouble finding a formal article for why this is normally done, but it's described here as [verb tense consistency](https://www.grammarly.com/blog/verb-tense-consistency/).

File: packages/in_app_purchase/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart [hunk 4]
```diff
@@ -3,12 +3,280 @@
...
+  final String transactionID;
+
+  /// The download progress, between 0.0 to 1.0.
+  final double progress;
+
+  /// The estimated time remaining for the download; if no good estimate is able to be made,
+  /// [downloadTimeUnknown] will be set to true.
+  final double timeRemaining;
+
+  /// true if [timeRemaining] cannot be estimated.
```
Comment (mklim) [inline packages/in_app_purchase/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart]: nit: "True"

File: packages/in_app_purchase/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart [hunk 5]
```diff
@@ -3,12 +3,280 @@
...
+/// Used as the parameter to initiate a payment.
+/// In general, a developer should not need to create the payment object explicitly; instead, use
+/// [SKPaymentQueueWrapper.addPayment] directly with a product identifier to initiate a payment.
+@JsonSerializable()
+class SKPaymentWrapper {
+  SKPaymentWrapper(
+      {@required this.productIdentifier,
+      @required this.applicationUsername,
+      this.requestData,
+      this.quantity = 1,
```
Comment (mklim) [inline packages/in_app_purchase/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart]: I think it's worth documenting that quantity defaults to 1 if unspecified.

File: packages/in_app_purchase/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart (hunk 3 above)
Comment (cyanglaz) [reply packages/in_app_purchase/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart]: going to rephrase the entire paragraph.

Comment (cyanglaz) [reply packages/in_app_purchase/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart]: be added is the correct way to do it, I actually also found it out myself while I'm reading it myself.

File: packages/in_app_purchase/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart (hunk 5 above)
Comment (cyanglaz) [reply packages/in_app_purchase/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart]: This is done on the property itself, do you mean to add a dart doc here as well? Or should we just have a comment?

File: packages/in_app_purchase/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart [hunk 6]
```diff
@@ -10,6 +10,9 @@ import 'package:json_annotation/json_annotation.dart';
 part 'sk_payment_queue_wrapper.g.dart';
 
 /// A wrapper around [`SKPaymentQueue`](https://developer.apple.com/documentation/storekit/skpaymentqueue?language=objc).
+///
+/// The payment queue contains payment related operations. It communicates with App Store and presents
+/// a user interface for user to process and authorize the payment.
```
Comment (mklim) [inline packages/in_app_purchase/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart]: nit: "for user" either needs a definite article ("for the user") or to be plural ("for users").

File: packages/in_app_purchase/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart [hunk 7]
```diff
@@ -105,6 +113,8 @@ class SKPaymentTransactionWrapper {
 }
 
 /// Dart wrapper around StoreKit's [SKDownloadState](https://developer.apple.com/documentation/storekit/skdownloadstate?language=objc).
+///
+/// The state of a download operation that can be in.
```
Comment (mklim) [inline packages/in_app_purchase/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart]: nit: remove "of", so "The state a download operation..."

File: packages/in_app_purchase/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart [hunk 8]
```diff
@@ -134,12 +144,12 @@ enum SKDownloadState {
 /// Dart wrapper around StoreKit's [SKDownload](https://developer.apple.com/documentation/storekit/skdownload?language=objc).
 ///
 /// When a product is created in the App Store Connect, one or more download contents can be associated with it.
-/// When the product is purchased, a List of download object will present in an [SKPaymentTransactionWrapper] object.
-/// Then download objects can be added to the payment queue.
-/// Read the [contentURL] to get the URL of the downloaded content after download complete.A
-/// All downloaded files must be processed before the completion of the transaction.
-/// After the transaction is complete, any download object in the transaction will not be able to add to the payment queue
-/// and the contentURL of the download object will be invalid.
+/// When the product is purchased, a List of [SKDownloadWrapper] object will present in an [SKPaymentTransactionWrapper] object.
```
Comment (mklim) [inline packages/in_app_purchase/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart]: nit: "[SKDownloadWrapper] objects will be present"

**Comment by cyanglaz** [comment]
Fixed the last nit comments, will merge in on green
